src/lib/optimizer/strategy/expression_reduction_rule.cpp

```diff
@@ -14,48 +16,38 @@ namespace opossum {
 
 using namespace opossum::expression_functional;  // NOLINT
 
-std::string LogicalReductionRule::name() const { return "Logical Expression Reducer Rule"; }
+std::string ExpressionReductionRule::name() const { return "Expression Reduction Rule"; }
 
-void LogicalReductionRule::apply_to(const std::shared_ptr<AbstractLQPNode>& node) const {
-  Assert(node->type == LQPNodeType::Root, "LogicalReductionRule needs root to hold onto");
-
-  /**
-   * Step 1:
-   *    - Reduce Expressions in the LQP
-   *    - Collect PredicateNodes that can be split up into multiple ones into `predicate_nodes_to_flat_conjunctions`
-   */
-  auto predicate_nodes_to_flat_conjunctions =
-      std::vector<std::pair<std::shared_ptr<PredicateNode>, std::vector<std::shared_ptr<AbstractExpression>>>>{};
+void ExpressionReductionRule::apply_to(const std::shared_ptr<AbstractLQPNode>& node) const {
+  Assert(node->type == LQPNodeType::Root, "ExpressionReductionRule needs root to hold onto");
 
   visit_lqp(node, [&](const auto& sub_node) {
-    // We only aim at PredicateNodes, since these are the nodes that primarily contain logical expressions.
-    if (const auto predicate_node = std::dynamic_pointer_cast<PredicateNode>(sub_node)) {
-      const auto new_predicate = reduce_distributivity(predicate_node->predicate());
-      const auto flat_conjunction = flatten_logical_expressions(new_predicate, LogicalOperator::And);
-
-      if (flat_conjunction.size() > 1) {
-        predicate_nodes_to_flat_conjunctions.emplace_back(predicate_node, flat_conjunction);
+    for (auto& expression : sub_node->node_expressions) {
+      // TODO(anybody) We're not doing a deep traversal of the Expressions here. The top-level expression must match
+      //               the patterns the techniques are looking for, otherwise no rewrite happens.
+      // TODO(anybody) Unnecessary shared_ptr copies being made here. Avoiding them is tricky: The expression
+      //               shared_ptr might have multiple owners, so we cannot rewrite in-place...
+
+      expression = reduce_distributivity(expression);
+      expression = reduce_in_with_single_list_element(expression);
+
+      // We can't prune Aggregate arguments, because the operator doesn't support, e.g., `MIN(1)`, whereas it supports
+      // `MIN(2-1)`, since `2-1` becomes a column.
+      if (sub_node->type != LQPNodeType::Aggregate) {
+        // TODO(anybody) We can't prune top level expressions right now, because that breaks `SELECT MIN(1+2)...`
+        //               because if we rewrite that to `SELECT MIN(3)...` the input to the aggregate is not a column
+        //               anymore and the Aggregate operator cannot handle that
+        for (auto& argument : expression->arguments) {
+          reduce_constant_expression(argument);
```

Comment:
Why do we modify in-place here but use return values in `reduce_in_with_single_list_element`?

Reply:
okay changed to in-place rewrite and also recursive application.

Tried this before but failed. probably just been tired.